concrete/dense_directed_matrix.go

```diff
@@ -112,8 +115,16 @@ func (g *DirectedDenseGraph) HasEdgeFromTo(u, v graph.Node) bool {
 	return uid != vid && !isSame(g.mat.At(uid, vid), g.absent)
 }
 
-func (g *DirectedDenseGraph) Weight(e graph.Edge) float64 {
-	return g.mat.At(e.From().ID(), e.To().ID())
+func (g *DirectedDenseGraph) Weight(x, y graph.Node) (w float64, ok bool) {
+	xid := x.ID()
+	yid := y.ID()
+	if xid == yid {
+		return 0, true
```

Comment:
>Having two parameters self and absent does not seem so bad if we consider correlation graphs, because now the code always returns 0. But calling Weight() with self edges is easily avoided, so we can postpone including g.self until it is clearer that it is really useful.

I'll add it. I've remembered that I need to add the absent param to some of the other graphs as well, so I'll do it all together when I respond to you comments here.